packages/minter-contracts/test-hs/Test/EnglishAuction/TezFixedFee.hs

```diff
@@ -0,0 +1,696 @@
+module Test.EnglishAuction.TezFixedFee where
+
+import Cleveland.Util (sec)
+import qualified Data.List as List
+import qualified Data.Map as Map
+import Data.Traversable.WithIndex (TraversableWithIndex, ifor)
+import Fmt ((+|), (|+))
+import Hedgehog (Gen, Property, forAll, label, property)
+import qualified Hedgehog.Gen as Gen
+import qualified Hedgehog.Range as Range
+import qualified Indigo.Contracts.FA2Sample as FA2
+import Lorentz hiding (amount, contract, now)
+import Lorentz.Contracts.EnglishAuction.TezFixedFee
+import Lorentz.Contracts.MinterSdk
+import Lorentz.Contracts.Spec.FA2Interface (TokenId(..))
+import Michelson.Typed (convertContract, untypeValue)
+import Morley.Nettest
+import Test.Util (balanceOf, clevelandProp, genMutez')
+import Tezos.Core (timestampPlusSeconds)
+
+hprop_Auctions_must_start_within_the_configured_number_of_seconds :: Property
+hprop_Auctions_must_start_within_the_configured_number_of_seconds =
+  property $ do
+    testData@TestData{testMaxConfigToStartTime} <- forAll genTestData
+
+    testTimeToStart <- forAll $ Gen.integral $
+      Range.linear
+        (toInteger testMaxConfigToStartTime + 1)
+        ((toInteger testMaxConfigToStartTime + 1 ) * 2)
+    let testData' = testData { testTimeToStart = testTimeToStart }
+
+    clevelandProp $ do
+      setup@Setup{contract} <- testSetup testData'
+
+      configureAuction testData' setup
+        `expectFailure`
+          failedWith contract [mt|start_time must not be greater than the sum of current time and max_config_to_start_time|]
+
+hprop_First_bid_is_valid_IFF_it_meets_opening_price :: Property
+hprop_First_bid_is_valid_IFF_it_meets_opening_price =
+  property $ do
+
+    testData@TestData{testOpeningPrice} <- forAll genTestData
+    firstBid <- forAll $ genMutez' (Range.linear 0 (testOpeningPrice * 2))
+
+    clevelandProp $ do
+      setup@Setup{seller, startTime, contract} <- testSetup testData
+      bidder <- newAddress "bidder"
+      withSender seller $ configureAuction testData setup
+      waitForAuctionToStart testData
+
+      withSender bidder $ do
+        if firstBid >= testOpeningPrice
+          then placeBid contract firstBid
+          else placeBid contract firstBid `expectFailure` failedWith contract
+                 ([mt|Invalid Bid amount|] :: MText, ((testOpeningPrice, firstBid) , seller, startTime), startTime)
+
+hprop_Subsequent_bids_are_valid_if_they_are_above_'min_raise' :: Property
+hprop_Subsequent_bids_are_valid_if_they_are_above_'min_raise' =
+  property $ do
+    testData@TestData{testOpeningPrice} <- forAll genTestData
+
+    firstBid <- forAll $ genBid testData testOpeningPrice
+
+    let minRaisePercentOfFirstBid = minRaisePercentOfBid testData firstBid
+    testMinRaise <-
+      forAll $ Gen.choice
+        [ genMutez' (Range.linear 1 minRaisePercentOfFirstBid)
+        , genMutez' (Range.linear minRaisePercentOfFirstBid (minRaisePercentOfFirstBid * 2))
```

Comment:
GREAT question, it made me think quite a lot to come up with an answer :sweat_smile:

Strap in! 

`Range.linear 1 (minRaisePercentOfFirstBid * 2)` does not mean there's a 50% chance of it being below or above `minRaisePercentOfFirstBid`. It means there will be a ~0% chance at the beginning, and _progressively_ get closer to a 50% chance, towards the end.

To see how this would work out in practice, let us take a few example runs. Hedgehog will run 100 tests, so let's analyse runs number 1, 25, 50, 75 and 100.
Assume `firstBid=1000`.

## Run no.1 (test size = 0)

`min_raise_percent` has a linear range [1, 100].
Because test size=0, the range's actual bounds will be [1, 1].
On average, `min_raise_percent` will be ~1.

Therefore, `minRaisePercentOfFirstBid = 1% of 1000 = 10`

Now, `min_raise` has a linear range of [1, minRaisePercentOfFirstBid*2] (so, [1,20])
Because test size=0, the range's actual bounds will be [1, 1].
On average, `min_raise` will be ~1.

`min_raise` is *very likely* to be below `minRaisePercentOfFirstBid`

## Run no.25 (test size = 24)

`min_raise_percent` has a linear range [1, 100].
Because test size=24, the range's actual bounds will be [1, 25].
On average, `min_raise_percent` will be ~12.5.

Therefore, `minRaisePercentOfFirstBid ~= 12.5% of 1000 ~= 125`

Now, `min_raise` has a linear range of [1, minRaisePercentOfFirstBid*2] (so, [1,250])
Because test size=24, the range's actual bounds will be [1, 62.5].
On average, `min_raise` will be ~31.25.

`min_raise` is *very likely* to be below `minRaisePercentOfFirstBid`

## Run no.50 (test size = 49)

`min_raise_percent` has a linear range [1, 100].
Because test size=49, the range's actual bounds will be [1, 50].
On average, `min_raise_percent` will be ~25.

Therefore, `minRaisePercentOfFirstBid ~= 25% of 1000 ~= 250`

Now, `min_raise` has a linear range of [1, minRaisePercentOfFirstBid*2] (so, [1,500])
Because test size=49, the range's actual bounds will be [1, 250].
On average, `min_raise` will be ~125.

`min_raise` is *very likely* to be below `minRaisePercentOfFirstBid`


## Run no.75 (test size = 74)

`min_raise_percent` has a linear range [1, 100].
Because test size=74, the range's actual bounds will be [1, 75].
On average, `min_raise_percent` will be ~37.5.

Therefore, `minRaisePercentOfFirstBid ~= 37.5% of 1000 ~= 375`

Now, `min_raise` has a linear range of [1, minRaisePercentOfFirstBid*2] (so, [1,750])
Because test size=74, the range's actual bounds will be [1, 562.5].
On average, `min_raise` will be ~281.25.

`min_raise` is *very likely* to be below `minRaisePercentOfFirstBid`

## Run no.100 (test size = 99)

`min_raise_percent` has a linear range [1, 100].
Because test size=99, the range's actual bounds will be [1, 100].
On average, `min_raise_percent` will be ~50.

Therefore, `minRaisePercentOfFirstBid ~= 50% of 1000 ~= 500`

Now, `min_raise` has a linear range of [1, minRaisePercentOfFirstBid*2] (so, [1,1000])
Because test size=99, the range's actual bounds will be [1, 1000].
On average, `min_raise` will be ~500.

`min_raise` has a *50% chance* of being below `minRaisePercentOfFirstBid` and 50% of being above.

-----------------


With `Gen.choice [ ... Range.linear (1, minRaisePercent), ... Range.linear minRaisePercent (minRaisePercent * 2) ]`, then the generator has a 50% chance of generating a min_raise above or below min_raise_percent _at each and every run_.

-----------------


A possible alternative would be to use `Range.constant 1 (minRaisePercentOfFirstBid * 2)` (instead of linear). But a big constant range (Say `Range.constant 0 1000`) means there's a good chance edge-case values (like 0 or 1) will probably not be tested for. And since bugs are usually found in the lower edges, it's important to prefer linear over constant ranges.

So, `Gen.choice`+2 `Range.linear`, when compared against a single `Range.constant`, has the advantage of ensuring good coverage of:
* the case where `min_raise` is really low
* the case where `min_raise` is really close to `minRaisePercentOfFirstBid`

Reply:
Oh yes, now it makes perfect sense. So one has to keep in mind that all the generators also have some specific distribution. And generalizing, in this case, `linear (1, 2 * X)` is semantically not what we actually wanted to generate if we take the distributions into account.

A bit sad that this seems to belong to some advanced best practices evaluated solely from experience or intuition, I didn't see this documented in the library. And someone reading this code may get confused as well.

Reply:
Indeed :confused: Ranges/bounds are really tricky to get right in more complex prop-based tests. Luckily, hedgehog gives me full control over these things (which is one of the reasons why I prefer it over quickcheck).

Also, I tend to use `collect` and `label` a lot to make sure the generators are evenly distributed. Sometimes I delete `collect`/`label` before committing, but I think in this case it will help in maintaining these tests in the future, so I left it in.

Reply:
`collect`/`label` is really a rescue for this. Looking forward to seeing a day when it will be possible to leave a check on them (so that test fails if too few of the expected samples were generated constantly) without worrying that CI will fail once in a week :smile: